modules/core/src/main/scala/org/tesselation/infrastructure/trust/TrustDaemon.scala

```diff
@@ -0,0 +1,74 @@
+package org.tesselation.infrastructure.trust
+
+import cats.Parallel
+import cats.effect.std.Random
+import cats.effect.{Async, Spawn, Temporal}
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tesselation.config.types.TrustDaemonConfig
+import org.tesselation.domain.Daemon
+import org.tesselation.domain.trust.storage.TrustStorage
+import org.tesselation.kryo.KryoSerializer
+import org.tesselation.schema.peer.PeerId
+import org.tesselation.security.SecurityProvider
+
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait TrustDaemon[F[_]] extends Daemon[F]
+
+object TrustDaemon {
+
+  def make[F[_]: Async: SecurityProvider: KryoSerializer: Random: Parallel](
+    cfg: TrustDaemonConfig,
+    trustStorage: TrustStorage[F],
+    selfPeerId: PeerId
+  ): TrustDaemon[F] = new TrustDaemon[F] {
+
+    private val logger = Slf4jLogger.getLogger[F]
+
+    def start: F[Unit] =
+      for {
+        _ <- Spawn[F].start(modelUpdate.foreverM).void
+      } yield ()
+
+    private def calculateTrust(): F[Unit] = {
+      val predicted = trustStorage.getTrust.map { trust =>
+        val fixed = trust.toList
+        val selfTrust = fixed.flatMap { case (peerId, trustInfo) => trustInfo.publicTrust.map(t => peerId -> t) }
+        val allNodes = fixed.map {
+          case (peerId, trustInfo) =>
+            peerId -> trustInfo.peerLabels
+        } :+ (selfPeerId -> selfTrust.toMap)
+        val peerIdToIdx = allNodes.zipWithIndex.map { case ((peerId, _), idx) => peerId -> idx }.toMap
+        val idxToPeerId = peerIdToIdx.map { case (k, v) => v -> k }
```

Comment:
Converting `Map` to `List` seems redundant because you do this in the calculation flow:

`Map[k, v]` -> `List[(k, v)]` -> `Map[k, v]` -> `List[k, v]` -> `Map[k,v]`

but we could do it without intermediate conversions:

`Map[k,v]` -> List[(k, v)] -> `Map[k, v]`

Also, you could use `mapValues` in some places as key (`peerId`) stays the same. That being said whole pure function for calculating predicted trust (see my other comments) could look like this:

```scala
        val selfTrustLabels = trust.flatMap { case (peerId, trustInfo) => trustInfo.publicTrust.map(peerId -> _) } // Map
        val allNodesTrustLabels = trust.view.mapValues(_.peerLabels).toMap + (selfPeerId -> selfTrustLabels) // Map

        val peerIdToIdx = allNodesTrustLabels.keys.zipWithIndex.toMap // Map
        val idxToPeerId = peerIdToIdx.map(_.swap) // Map

        val trustNodes = allNodesTrustLabels.map {
          case (peerId, labels) =>
            TrustNode(peerIdToIdx(peerId), 0, 0, labels.map {
              case (pid, label) =>
                TrustEdge(peerIdToIdx(peerId), peerIdToIdx(pid), label, peerId == selfPeerId)
            }.toList)
        }.toList // Map -> List because SAW needs List
        
        SelfAvoidingWalk
          .runWalkFeedbackUpdateSingleNode(peerIdToIdx(selfPeerId), trustNodes)
          .edges
          .map(e => idxToPeerId(e.dst) -> e.trust)
          .toMap // List -> Map
```

Reply:
Fixed